### PR TITLE
[BE][inductor] Type annotate `codecache.py` and `config.py`

### PR DESCRIPTION
**Pull request by YUNQIUGUO**
As title. 

Checked/ Referred to the raw json file for runtime types . (and tried to cover all the missing annotations listed in the .json) this time.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -448,7 +449,8 @@ def _ident(x: Any) -> Any:
     return x
 
 
-def extract_tensor_metadata_for_cache_key(device_map, t):
+def extract_tensor_metadata_for_cache_key(device_map: Dict[torch.device, torch.device],
```
Comment (eellison) [inline torch/_inductor/codecache.py:452]: nit: `FakeTensor` subtypes `Tensor` so no need for Union

File: torch/_inductor/codecache.py [hunk 2]
```diff
@@ -471,15 +473,17 @@ def extract_tensor_metadata_for_cache_key(device_map, t):
     return meta
 
 
-def _reduce_fake_tensor(device_map, t):
+def _reduce_fake_tensor(device_map: Dict[torch.device, torch.device],
+                        t: FakeTensor) -> Tuple[Callable[..., Any], Tuple[TensorMetadata]]:
```
Comment (eellison) [inline torch/_inductor/codecache.py:477]: The Callables are tricky, generally. in this case, `_ident` is just :

T = TypeVar('T')

`Callable[[T], T]`

Would you mind updating this and the _ident declaration below?

File: torch/_inductor/codecache.py [hunk 3]
```diff
@@ -471,15 +473,17 @@ def extract_tensor_metadata_for_cache_key(device_map, t):
...
     """
     See FxGraphCachePickler. Custom reducer to pickle FakeTensors.
     """
     metadata = extract_tensor_metadata_for_cache_key(device_map, t)
     return (_ident, (metadata,))
 
 
-def _reduce_tensor(device_map, t):
+def _reduce_tensor(device_map: Dict[torch.device, torch.device],
+                   t: Tensor) -> Tuple[Callable[..., Any], Tuple[TensorMetadataAndValues]]:
```
Comment (eellison) [inline torch/_inductor/codecache.py:486]: same comment about ident

File: torch/_inductor/codecache.py [hunk 4]
```diff
@@ -510,7 +514,7 @@ def _reduce_tensor(device_map, t):
     return (_ident, (TensorMetadataAndValues(metadata, values),))
 
 
-def _reduce_symint(s):
+def _reduce_symint(s: SymInt) -> Tuple[Callable[..., Any], Tuple[str]]:
```
Comment (eellison) [inline torch/_inductor/codecache.py:517]: and here

**Review by eellison — APPROVED** [review]
Looks good ! just a few comments.

**Comment by oulgen** [comment]
Do we not want to remove "# mypy: allow-untyped-defs" from top of the file?

**Comment by eellison** [comment]
@oulgen oops yea good catch

**Comment by YUNQIUGUO** [comment]
> Do we not want to remove "# mypy: allow-untyped-defs" from top of the file?

Hi folks, thanks for mentioning this! 

For `codecache.py` there's still a couple functions not type annotated. (and they are not listed in the raw json files)
I am not sure if it's worth adding type annotations though:

To name a few:

```
 def log_global_cache_stats(*args, **kwargs):
         pass
```
```
 def log_global_cache_errors(*args, **kwargs):
       pass
```
...

```
    >>> 3270  |    def __exit__(self, *args):
        3271  |        self.close()
```

So I still kept the `mypy` thing at the top. Let me know what you think

**Comment by eellison** [comment]
@YUNQIUGUO it's okay to have particular functions where type annotations are not that useful, and you can add `Any` annotations or ignrore specific functions as needed.

But it's still useful to remove the ignore from the top of the file because it enforces that new functions that we want annotated will be typed

**Comment by YUNQIUGUO** [comment]
> @YUNQIUGUO it's okay to have particular functions where type annotations are not that useful, and you can add `Any` annotations or ignrore specific functions as needed.
> 
> But it's still useful to remove the ignore from the top of the file because it enforces that new functions that we want annotated will be typed

ok I see! thanks for the explanation. that makes sense for me. will update

**Review by oulgen — APPROVED** [review]
Thanks for the type annotations!

**Pull request closed without merging**